### ISSUE TITLE
Race condition utility.slug

### ISSUE THREAD

**Issue by supakeen** [supakeen/pinnwand#34]
It seems that when there is few slugs in the database that the utility function that creates slugs can in fact generate a slug that already exists in the database.

I've seen this happen once just after a redeploy. Perhaps we want to catch the IntegrityError and try again or look into why this happens.

Happens on the commit in CreateAction.

**Comment by supakeen** [comment]
This issue seems to still be active.

**Comment by mweinelt** [comment]
```
Jun 20 12:42:06 juno pinnwand[5271]: ERROR:tornado.application:Uncaught exception POST /create (2003:d4:bf45:1102:55e6:c8e2:bfbf:3533)
Jun 20 12:42:06 juno pinnwand[5271]: HTTPServerRequest(protocol='https', host='paste.lossy.network', method='POST', uri='/create', version='HTTP/1.0', remote_ip='2003:d4:bf45:1102:55e6:c8e2:bfbf:3533')
Jun 20 12:42:06 juno pinnwand[5271]: Traceback (most recent call last):
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1283, in _execute_context
Jun 20 12:42:06 juno pinnwand[5271]:     self.dialect.do_execute(
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 590, in do_execute
Jun 20 12:42:06 juno pinnwand[5271]:     cursor.execute(statement, parameters)
Jun 20 12:42:06 juno pinnwand[5271]: sqlite3.IntegrityError: UNIQUE constraint failed: file.slug
Jun 20 12:42:06 juno pinnwand[5271]: The above exception was the direct cause of the following exception:
Jun 20 12:42:06 juno pinnwand[5271]: Traceback (most recent call last):
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/cbidsdgi32mr4fyhsiqfb2kl7cnxs6sw-python3.8-tornado-6.0.4/lib/python3.8/site-packages/tornado/web.py", line 1701, in _execute
Jun 20 12:42:06 juno pinnwand[5271]:     result = method(*self.path_args, **self.path_kwargs)
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/vsn5am62lha3jb6684g53n2ryz4f8v9j-pinnwand-1.1.2/lib/python3.8/site-packages/pinnwand/handler/website.py", line 197, in post
Jun 20 12:42:06 juno pinnwand[5271]:     session.commit()
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 1042, in commit
Jun 20 12:42:06 juno pinnwand[5271]:     self.transaction.commit()
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 504, in commit
Jun 20 12:42:06 juno pinnwand[5271]:     self._prepare_impl()
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 483, in _prepare_impl
Jun 20 12:42:06 juno pinnwand[5271]:     self.session.flush()
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 2523, in flush
Jun 20 12:42:06 juno pinnwand[5271]:     self._flush(objects)
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 2664, in _flush
Jun 20 12:42:06 juno pinnwand[5271]:     transaction.rollback(_capture_exception=True)
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
Jun 20 12:42:06 juno pinnwand[5271]:     compat.raise_(
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 178, in raise_
Jun 20 12:42:06 juno pinnwand[5271]:     raise exception
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 2624, in _flush
Jun 20 12:42:06 juno pinnwand[5271]:     flush_context.execute()
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/orm/unitofwork.py", line 422, in execute
Jun 20 12:42:06 juno pinnwand[5271]:     rec.execute(self)
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/orm/unitofwork.py", line 586, in execute
Jun 20 12:42:06 juno pinnwand[5271]:     persistence.save_obj(
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/orm/persistence.py", line 239, in save_obj
Jun 20 12:42:06 juno pinnwand[5271]:     _emit_insert_statements(
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/orm/persistence.py", line 1135, in _emit_insert_statements
Jun 20 12:42:06 juno pinnwand[5271]:     result = cached_connections[connection].execute(
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1020, in execute
Jun 20 12:42:06 juno pinnwand[5271]:     return meth(self, multiparams, params)
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/sql/elements.py", line 298, in _execute_on_connection
Jun 20 12:42:06 juno pinnwand[5271]:     return connection._execute_clauseelement(self, multiparams, params)
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1133, in _execute_clauseelement
Jun 20 12:42:06 juno pinnwand[5271]:     ret = self._execute_context(
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1323, in _execute_context
Jun 20 12:42:06 juno pinnwand[5271]:     self._handle_dbapi_exception(
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1517, in _handle_dbapi_exception
Jun 20 12:42:06 juno pinnwand[5271]:     util.raise_(
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 178, in raise_
Jun 20 12:42:06 juno pinnwand[5271]:     raise exception
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1283, in _execute_context
Jun 20 12:42:06 juno pinnwand[5271]:     self.dialect.do_execute(
Jun 20 12:42:06 juno pinnwand[5271]:   File "/nix/store/y4vrky9xj4lmgk0avldv26dikn7wxs76-python3.8-SQLAlchemy-1.3.17/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 590, in do_execute
Jun 20 12:42:06 juno pinnwand[5271]:     cursor.execute(statement, parameters)
Jun 20 12:42:06 juno pinnwand[5271]: sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: file.slug
Jun 20 12:42:06 juno pinnwand[5271]: [SQL: INSERT INTO file (paste_id, slug, pub_date, chg_date, lexer, raw, fmt, filename) VALUES (?, ?, ?, ?, ?, ?, ?, ?)]
Jun 20 12:42:06 juno pinnwand[5271]: [parameters: (26, 'AE', '2020-06-20 10:42:06.043578', '2020-06-20 10:42:06.043597', 'nixos', '{ config, pkgs, ... }:\r\nlet\r\n  riotWebConfig = {\r\n    default_server_config = {\r\n      "m.homeserver" = {\r\n        base_url = "https://matr ... (2030 characters truncated) ... eSSL = true;\r\n    useACMEHost = "${config.networking.hostName}.${config.networking.domain}";\r\n\r\n    root = "${pkgs.riot-web}";\r\n  };\r\n}\r\n', '<div class="source"><table class="sourcetable"><tr><td class="linenos linenodiv"><code data-line-number=" 1"></code></td><td class="code"><code><span ... (16875 characters truncated) ... td class="linenos linenodiv"><code data-line-number="74"></code></td><td class="code"><code><span class="p">}</span>\n</code></td></tr></table></div>', 'riot.nix')]
Jun 20 12:42:06 juno pinnwand[5271]: (Background on this error at: http://sqlalche.me/e/gkpj)
Jun 20 12:42:06 juno pinnwand[5271]: ERROR:tornado.access:500 POST /create (2003:d4:bf45:1102:55e6:c8e2:bfbf:3533) 101.19ms
Jun 20 12:42:17 juno pinnwand[5271]: INFO:tornado.access:302 POST /create (2003:d4:bf45:1102:55e6:c8e2:bfbf:3533) 354.01ms
Jun 20 12:42:17 juno pinnwand[5271]: INFO:tornado.access:200 GET /LA (2003:d4:bf45:1102:55e6:c8e2:bfbf:3533) 38.10ms
```

**Comment by supakeen** [comment]
This occurs due to the way slugs are created, they exist in the session but this is not enforced by sqlalchemy (nor are we passing the session around).

I just tried passing the session to slug_create which also doesn't solve the issue due to again, sqlalchemy not enforcing uniqueness there.

The solution would be to have a 'slug creation' state of the currently generated slugs and adds them to the already taken slugs for slug_create. Perhaps this can be done with a context manager.

**Comment by supakeen** [comment]
With the current fix while the issue has been reduced there are still intermittent errors, more code is necessary the errors only occur in the API testcase.

**Comment by supakeen** [comment]
This condition is finally fixed and v1.1.3 is released with the fix.
